src/graph.ts

```diff
@@ -11,35 +11,148 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
 import { debug } from '@pulumi/pulumi/log';
-import { Stack, CfnElement, Token } from 'aws-cdk-lib';
-import { Construct, ConstructOrder } from 'constructs';
-import { CloudFormationTemplate } from './cfn';
+import { CloudFormationResource } from './cfn';
 import { parseSub } from './sub';
+import { ConstructTree, StackManifest } from './assembly';
+
+/**
+ * Represents a CDK Construct
+ */
+export interface ConstructInfo {
+    /**
+     * The construct path
+     */
+    path: string;
+
+    /**
+     * The node id of the construct
+     */
+    id: string;
+
+    /**
+     * The CloudFormation resource type
+     *
+     * This will only be set if this is the construct for cfn resource
+     */
+    type?: string;
+
+    /**
+     * The attributes of the construct
+     */
+    attributes?: { [key: string]: any };
+
+    /**
+     * The parent construct (i.e. scope)
+     * Will be undefined for the construct representing the `Stack`
+     */
+    parent?: ConstructInfo;
+}
 
 export interface GraphNode {
     incomingEdges: Set<GraphNode>;
     outgoingEdges: Set<GraphNode>;
-    construct: Construct;
-    template?: CloudFormationTemplate;
+    /**
+     * The CFN LogicalID.
+     *
+     * This will only be set if this node represents a CloudFormation resource.
+     * It will not be set for wrapper constructs
+     */
+    logicalId?: string;
```

Comment:
Worth commenting what incoming and outgoing edges are? Something with dependency information? I was not super sure. What does it mean to have an outgoingEdge?

Also, `Set<GraphNode>` compares on hash equality? SO this is what we want here, one node = one in-mem object?